src/pagination.rs

```diff
@@ -0,0 +1,273 @@
+use std::cmp::Ordering;
+use std::ops::Range;
+use std::path::PathBuf;
+
+use liquid;
+
+use cobalt_model::pagination_config::Include;
+use cobalt_model::pagination_config::PaginationConfig;
+use cobalt_model::permalink;
+use cobalt_model::SortOrder;
+use document;
+use document::Document;
+
+pub fn generate_paginators(
+  doc: &mut Document,
+  posts_data: &[liquid::Value],
+) -> Vec<liquid::Object> {
+  let config = doc.front.pagination.as_ref().unwrap();
+  let mut all_posts = posts_data.to_vec();
+  match config.include {
+    Include::All => {
+      sort_posts(&mut all_posts, &config);
+      create_all_paginators(&mut all_posts, &doc, &config)
+    }
+    Include::None => Vec::new(),
+  }
+}
+
+fn extract_page_path_from(p: &liquid::Object) -> String {
+  p.get("page_path")
+    .expect("Should have a page_path")
+    .as_scalar()
+    .expect("Should be a scalar")
+    .to_str()
+    .into_owned()
+}
+
+pub fn extract_page_path(p: &liquid::Object, config: &PaginationConfig) -> PathBuf {
+  PathBuf::from(match config.include {
+    Include::All => extract_page_path_from(&p),
+    _ => "".to_owned(),
+  })
+}
+
+fn extract_value(a: &liquid::Value, key: &String) -> Option<liquid::Scalar> {
+  a.get(&key.clone().into())
+    .map_or(None, |sort_key| sort_key.as_scalar().cloned())
+}
+
+// sort posts by multiple criteria
+fn sort_posts(posts: &mut Vec<liquid::Value>, config: &PaginationConfig) {
+  posts.sort_by(|a, b| {
+    let keys = &config.sort_by;
+    let mut cmp = Ordering::Less;
+    for k in keys {
+      cmp = if let Some(a) = extract_value(a, &k) {
+        if let Some(b) = extract_value(b, &k) {
+          match config.order {
+            SortOrder::Desc => b.partial_cmp(&a).unwrap_or(Ordering::Equal),
```

Comment:
Its a bit annoying having this match in the inner loop.  It makes the code harder to follow.

Maybe express these as lambdas?

```rust
let cmp = match config.order {
    SortOrder::Desc => |a, b| b.partial_cmp(&a).unwrap_or(Ordering::Equal),
    ...
}
```
(would require renaming `cmp` to `order` or something)